source/sermux-proto/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+//! # sermux-proto
+//!
+//! Wire types used by the `SerialMuxService` in the kernel. Extracted as a
+//! separate crate to allow external decoders (like `crowtty`) to share protocol
+//! definitions
+
+#![cfg_attr(not(any(test, feature = "use-std")), no_std)]
+
+use core::mem::size_of;
+
+////////////////////////////////////////////////////////////////////////////////
+// Well Known Ports
+////////////////////////////////////////////////////////////////////////////////
+
+/// Well known `SerialMuxService` ports
+#[repr(u16)]
+#[non_exhaustive]
+pub enum WellKnown {
+    /// A bidirectional loopback channel - echos all characters back
+    Loopback = 0,
+    /// An output-only channel for sending periodic sign of life messages
+    HelloWorld = 1,
+    /// An input-only channel to act as a keyboard for a GUI application
+    /// such as a forth console, when there is no hardware keyboard available
+    PsuedoKeyboard = 2,
+    /// A bidirectional for binary encoded tracing messages
+    BinaryTracing = 3,
+}
+
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum Error {
+    /// The provided buffer is not suitable in size
+    InsufficientSize,
+    /// Ran out of room while filling a buffer, this is likely
+    /// an error in the `sermux-proto` library.
+    UnexpectedBufferFull,
+    /// The cobs decoding process failed. The message was likely
+    /// malformed or not a sermux-proto frame
+    CobsDecodeFailed,
+    /// Cobs decoding succeeded, but the resulting data was not
+    /// a valid sermux-proto frame
+    MalformedFrame,
+}
```

Comment:
this isn't a huge deal, but: the `InsufficientSize` and `UnexpectedBufferFull` errors are returned only when encoding, and `CobsDecodeFailed` and `MalformedFrame` errors are returned only when encoding --- maybe it's worth splitting this type into separate `EncodeError` and `DecodeError` enums, so that when handling the result of an encode or decode operation, the user doesn't have to match on variants that are never returned by that operation?